### DISCUSSION TITLE
Use of `run-id` for pinning workflow to particular runner for hyper optimizing docker jobs

### DISCUSSION THREAD

**Discussion by ruffsl** [runs-on/runs-on#72] [Q&A]
From the installation guide, one particular sentence that caught my eye was the recommendation of the run-id label:

``` diff
- runs-on: ubuntu-latest
+ runs-on: runs-on,runner=2cpu-linux-x64,run-id=${{ github.run_id }}
```

> We recommend adding the `run-id` label so that the runner is pinned to this workflow run. Otherwise the job could be executed by any other RunsOn runner having the same labels, which could make debugging a bit harder if anything goes wrong.
- https://runs-on.com/guides/install/

In particular, one of my action workflows follows this formulaic pattern of building and running docker images:

- Build and push `base` image layers via `docker/bake-action@v4`
- Run job container from `base` image to compile and cache project artifacts
- Run job container from `base` image to test and validate project artifacts
- Build and push `prod` image ontime of `base` layers using restored project artifacts
- Run job container from `prod` image to simulated staging of production services 

In practice this chain of consecutive jobs may extend further in depth (or branch wider in concurrent width), however, for the main hot path of building the base image, compiling/testing artifacts, and pushing the prod image, one could shave off a lot of time by coordinating the hot jobs onto the same runner instance, thus avoiding the need to redownload image layers pre-cached on the runner's host docker daemon.

With caching project artifacts, some of these hot jobs are can be essentially NoOps, but for the sake of determinism and cache bust detection, they must be executed, even if they take less than 10sec to complete. E.g: `docker buildx bake` with a hot `cache-from` source. Thus, most of the job's time (for lower bound limits) can be chalked up to simply re-downloading GB of shared image layers for each new runner assigned to pick up the job.

For job containers at least, I believe other runner hosting options, like CircleCI or perhaps GitHub, leverage heuristics when scheduling/provisioning jobs to co-locate them with runners that have previously targeted similar docker image/tag, their by minimizing the amount of layers that must be fetched either from a container registry or other docker cache backends.

For GitHub actions, I suspect this only most effective if the user deliberately disableds any docker cleanup:

``` yaml
  steps:
    - name: Setup Docker Buildx
      uses: docker/setup-buildx-action@v3
      with:
        cleanup: false
```

Yet for deterministic execution, the images for my container jobs are specified via docker image digest ID, rather than by docker image tag name, so that consecutive container jobs run from precisely the same exact image a prior job builds/pushes, regardless of how the state of the registry tags are change between job startup times. Thus heuristics that naive operate on the docker image tag name alone can easily result in provisioning a runner with few commonly pre-cached layers. A more rigorous approach would be to query the container image registry for the image manifest that includes the layer digests themselves that could directly translate to a more robust cost function for scheduling heuristic purposes.

Short of all that, it would be simpler to enable the workflow author to express such constraints explicitly via this `run-id` label. Having not yet used RunsOn just yet, but having looked into parameterizing the `runs-on:` job input configuration at runtime for all jobs within a triggered workflow, would this be an accurate and feasible assessment of this feature?

Related:

- https://github.com/actions/actions-runner-controller/discussions/2944#discussioncomment-7686485

**Comment by crohr** [comment]
Hello, it would be interesting to see your job dependencies to better understand why there needs to be so many coordinated steps between them. 

From your description though I believe you would be well served by using the S3 cache backend for docker layers instead of pushing artefacts since it can effectively serve as a hot cache. 

https://runs-on.com/reference/caching/

I also wonder whether you could inline some of your steps in a single job.

In RunsOn the machines are all ephemeral and discarded after the job is completed, so there is no way to execute multiple jobs on the same machine. GitHub behaves exactly the same.

**Reply by ruffsl** [reply]
>  it would be interesting to see your job dependencies to better understand why there needs to be so many coordinated steps between them

For our development cycle, the building and pushing of base image layers includes approximately three docker bake targets: one for runtime dependencies, another on top of that with testing dependencies, and a final layer on top of that for build time dependencies:

```
.
└── base ~ 406MB
    └── runner ~ 3.78GB
        ├── tester ~ 3.93GB
        │   └── builder ~ 4.58GB
        └── releaser ~ 4.11GB
```

The `builder` image is used for compilation of project artifacts, while the `tester` is similarly used testing said compiled artifacts. The key being that the test environment excludes any build time dependencies to better catch any undeclared runtime dependencies via CI that developers may have overlooked via personal workstations. Later jobs may pick up from this by packaging the artifacts for production via a `releaser` image, that is simply based upon the `runner` image layers, as to minimize image size for efficient deployment.

The software stack I'm working with is a fairly broad ecosystem, rather than say a single platform or monolingual framework. For comparison, the sizes for each of these stages/images are noted above. So, to minimize the number of layers and maximize for caching, while still prioritizing a minimal yet verifiable production environment, the layer hierarchy above is the current result. As our feature development continues, I foresee these images sizes growing until we'll the luxury of time to subdivide our stack further into separate containers.

---

> I believe you would be well served by using the S3 cache backend for docker layers instead of pushing artefacts since it can effectively serve as a hot cache.

Enabling the S3 cache backend is indeed something I already have queued for experimentation, but regardless of the matter in using the registry as a `cache-from` source or not, pushing to the registry remains a prerequisite because:

- Subsequent jobs, particularly container jobs, would need to pull the container image in order to commence
  - especially if machines can only be ephemeral per job, rather than be extended to be ephemeral per workflow
- To provide developers access to the same exact docker image digest ID that was used in a CI workflow
  - for greater debugging and introspection, without having to dissect the environment on a live CI machine via ssh

---

> I also wonder whether you could inline some of your steps in a single job.

That was an initial hope, but given the following constraints, inlining everything into a monolithic job has drawbacks:

- Unlike other CI frameworks, GH actions (AFAIK) can only associate jobs with a container, not individual steps
  - Thus leveraging the same DX ergonomics of GH actions with manual container runs would be tedious
- Not all jobs for different workflows types may use the same runner machine type
  - Feature workflows for quickly building and running unit tests could just as well use the same instance type
  - However, integration workflows may need GPU access for testing/simulation, but not necessary for building

To keep our actions maintainable and DRY, I'd like to share as many reusable workflows as across feature and integration workflows as reasonably possible. This is currently quite viable by parameterizing the `runs-on:` input of subsequent jobs via output of prior `need:`ed jobs.

---

> machines are all ephemeral and discarded after the job is completed, so there is no way to execute multiple jobs on the same machine.

While that is currently the case for GitHub hosted runners, it needn't apply in general. I'm guessing you're more aware of prior approaches to self hosting github runners via AWS than I, but one such approach we've employed before includes using a GitHub hosted runner to manually call the AWS API to spawn an EC2 instance, add that instance as an available self hosted runner for the org, then pass the resulting runner label to the various sequence of multiple jobs, then remember to stop said instance prior to concluding the workflow run.

- https://github.com/machulav/ec2-github-runner
- https://github.com/machulav/ec2-github-runner?tab=readme-ov-file#example

Of course, RunsOn's GitHub App integration is soo much more elegant and preferable as a stand in replacement. I'm thinking a simple user definable flag in the RunOn configuration could be added to optionally forstall the discarding of a runner instance. enabling the opportunity for it to be picked up again for another job that includes an exact matching `run-id`. This could be a post-job-completion event to lookup the job queue and check for matching `run-id`s before terminating the existing runner. Or perhaps and/or a simple timeout duration, given some seconds of latency for GitHub to  request for the next job in the hot-path of a workflow.

I mention "optionally", given that some users probably always intend to allocate a job to a fresh/sterile runner, but as the use case above depicts, recycling runners for jobs (at least within the same workflow) can be desirable as well. I know this could cause issues for those relying on spot instances, but as per the RunsOn docs, our workflows currently run well under an hour with external caching available. And if even it spot terminated, the workflow is still ambidextrous for ephemeral machines (given the occasional need for running from different instance types in the same workflow); so the occasional rerun/resume from last-failed-job is not a real blocker for low priority workflows using spot instances anyhow.

**Comment by crohr** [comment]
Thanks for the explanations, I understand the need for proper image pushing/tagging since you are using job containers. In that case I guess the build-push-action to a local ECR repository with either S3 based cache or registry based cache is your best option…

Regarding the reuse of a runner in certain conditions, this could in theory work, but RunsOn uses one-time JIT registration, so this means we would need some way to generate a new token and pass it to the runner. For now I don’t think I will integrate that kind of complexity in the product. I think it would make sense to experiment with what RunsOn has to offer, and then see the real pain points. 

Maybe offering a transparent way to (efficiently) restore the `/var/lib/docker` across runs would be better.

**Comment by crohr** [comment]
@ruffsl would be great to have your feedback after a few months of usage here. Were you able to devise sufficiently fast job workflows around your constraints in the end?

**Comment by crohr** [comment]
I'm doing a bit of cleanup and going to close this discussion, please let me know if you have any update!
